sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala

```diff
@@ -19,274 +19,575 @@ package org.apache.spark.sql.execution.streaming
+  //
+  // If we were to use the other deleteFromTTLIndex method, we would have to re-encode the
+  // components into an UnsafeRow. It is more efficient to just pass the UnsafeRow that we
+  // read from the iterator.
+  protected def deleteFromTTLIndex(ttlKey: UnsafeRow): Unit = {
+    store.remove(ttlKey, TTL_INDEX)
+  }
+
+  private[sql] def toTTLRow(ttlKey: UnsafeRow): TTLRow = {
+    val expirationMs = ttlKey.getLong(0)
+    val elementKey = ttlKey.getStruct(1, TTL_INDEX_KEY_SCHEMA.length)
+    TTLRow(elementKey, expirationMs)
+  }
+
+  private[sql] def getTTLRows(): Iterator[TTLRow] = {
+    store.iterator(TTL_INDEX).map(kv => toTTLRow(kv.key))
+  }
+
+  // Returns an Iterator over all the keys in the TTL index that have expired. This method
+  // does not delete the keys from the TTL index; it is the responsibility of the caller
+  // to do so.
+  //
+  // The schema of the UnsafeRow returned by this iterator is (expirationMs, elementKey).
+  protected def ttlEvictionIterator(): Iterator[UnsafeRow] = {
+    val ttlIterator = store.iterator(TTL_INDEX)
+
+    // Recall that the format is (expirationMs, elementKey) -> TTL_EMPTY_VALUE_ROW, so
+    // kv.value doesn't ever need to be used.
+    ttlIterator.takeWhile(kv => {
+      val expirationMs = kv.key.getLong(0)
+      StateTTL.isExpired(expirationMs, batchTimestampMs)
+    }).map(_.key)
+  }
+
+
+  // Encapsulates a row stored in a TTL index.
+  protected case class TTLRow(elementKey: UnsafeRow, expirationMs: Long)
 
   /**
-   * Perform the user state clean up based on ttl values stored in
-   * this state. NOTE that its not safe to call this operation concurrently
-   * when the user can also modify the underlying State. Cleanup should be initiated
-   * after arbitrary state operations are completed by the user.
+   * Evicts the state associated with this stateful variable that has expired
+   * due to TTL. The eviction applies to all grouping keys, and to all indexes,
+   * primary or secondary.
+   *
+   * This method can be called at any time in the micro-batch execution,
+   * as long as it is allowed to complete before subsequent state operations are
+   * issued. Operations to the state variable should not be issued concurrently while
+   * this is running.
+   *
+   * (Why? Some cleanup operations leave the state store in an inconsistent state while
+   * they are doing cleanup. For example, they may choose to get an iterator over all
+   * of the existing values, delete all values, and then re-insert only the non-expired
+   * values from the iterator. If a get is issued after the delete happens but before the
+   * re-insertion completes, the get could return null even when the value does actually
+   * exist.)
    *
    * @return number of values cleaned up.
    */
-  def clearExpiredState(): Long
+  def clearExpiredStateForAllKeys(): Long
+
+  /**
+   * When a user calls clear() on a stateful variable, this method is invoked to
```

Comment:
Hm, I'm not sure about this. When I did a code review with @anishshri-db, he was initially confused at what the difference between `clearExpiredStateForAllKeys` and `clearAllStateForElementKey` was. One of his questions was, "when do we invoke this?" So, I tried to answer that here.

IMO the docs should give context; for actual method details; developers can read code if they want method-level implementation detail.

(I can agree that my comments might be verbose, and I can shorten them.)

Reply:
Honestly saying (I admit I'm bad at it too), it sounds like actually an issue of naming methods, not a documentation issue.

I think this trait is making unnecessary confusion because this does beyond handling TTL; it's not a pure mix-in to attach the TTL ability but takes more responsibility e.g. removing primary index from clearAllStateForElementKey. You can move the responsibility to derived class (we do that on updating data) and only deal with TTL metadata in this trait - after that, the term is not needed to be `State` which is too overloaded usage.

I don't have strong opinion of whether we should change the code comment; I've said it's a nit, so just consider this as 2 cents. But I might argue strongly if this were user facing API.